lib/sample_uploader/utils/importer.py

```diff
@@ -212,77 +237,133 @@ def import_samples_from_file(
+                val = columns_to_input_names.pop("material")
+                columns_to_input_names["SESAR:material"] = val
 
-    acls = {
-        "read": [],
-        "write": [],
-        "admin": [],
-        "public_read": -1  # set to false (<0)
-    }
-    if params.get('share_within_workspace'):
-        # query workspace for user permissions.
-        acls = get_workspace_user_perms(workspace_url, params.get('workspace_id'), token, username, acls)
-    groups = SAMP_SERV_CONFIG['validators']
+        acls = {
+            "read": [],
+            "write": [],
+            "admin": [],
+            "public_read": -1  # set to false (<0)
+        }
+        if params.get('share_within_workspace'):
+            # query workspace for user permissions.
+            acls = get_workspace_user_perms(workspace_url, params.get('workspace_id'), token, username, acls)
+        groups = SAMP_SERV_CONFIG['validators']
+
+        cols = list(set(df.columns) - set(REGULATED_COLS))
+        sample_url = get_sample_service_url(sw_url)
+        samples, existing_samples, produce_errors = _produce_samples(
+            df,
+            cols,
+            column_groups,
+            column_unit_regex,
+            sample_url,
+            token,
+            input_sample_set['samples'],
+            columns_to_input_names,
+            header_row_index
+        )
+        errors += produce_errors
+    
+    if params.get('prevalidate') and not errors:
+        error_detail = validate_samples([s['sample'] for s in samples], sample_url, token)
+        errors += [ SampleContentError(
+                e['message'],
+                sample_name=e['sample_name'], 
+                node=e['node'], 
+                key=e['key']
+            ) for e in error_detail ]
 
-    cols = list(set(df.columns) - set(REGULATED_COLS))
-    sample_url = get_sample_service_url(sw_url)
-    samples, existing_samples = _produce_samples(
-        df,
-        cols,
-        column_groups,
-        column_unit_regex,
-        sample_url,
-        token,
-        input_sample_set['samples'],
-        columns_to_input_names
-    )
-    errors = {}
-    if params.get('prevalidate'):
-        errors = validate_samples([s['sample'] for s in samples], sample_url, token)
     if errors:
         saved_samples = []
+        # Fill in missing location information for SamplesContentError(s)
+        err_col_keys = {}
+        err_key_indices = {}
+        for col_idx, col_name in enumerate(df.columns):
+            err_col_keys[col_idx] = col_name
+            err_key_indices[col_name] = col_idx
+            if col_name in columns_to_input_names and columns_to_input_names[col_name] != col_name:
+                err_key_indices[columns_to_input_names[col_name]] = col_idx
+
+        err_row_sample_names = {}
+        err_sample_name_indices = {}
+        for row_idx, row in df.iterrows():
+            row_pos = header_row_index + 1 + row_idx
```

Comment:
maybe explain the `+1` here

Reply:
I can add a comment, not sure how to phrase it concisely, here is a visual explination:
```
0
1
2
3 <-- header row
4 <-- first sample, dataframe index "0"
5 
6 <-- some sample, dataframe index "3"
```
knowing the header position (3 above) and dataframe index (starts at zero after the header),

header_position + 1 = first_sample_row
first_sample_row + target_data_frame_index = target_sample_row

actually, might just change the code to read more like the above

Reply:
ah I see, that makes sense. i think changing the code like you describe would be helpful for readability